server/application/application.go

```diff
@@ -1498,7 +1499,67 @@ func (s *Server) RevisionMetadata(ctx context.Context, q *application.RevisionMe
 		return nil, err
 	}
 
-	source := a.Spec.GetSource()
+	var versionId int64 = 0
+	if q.VersionId != nil {
+		versionId = int64(*q.VersionId)
+	}
+
+	var source *v1alpha1.ApplicationSource
+
+	// To support changes between single source and multi source revisions
+	// we have to calculate if the operation has to be done as multisource or not.
+	// There are 2 different scenarios, checking current revision and historic revision
+	// - Current revision (VersionId is nil or 0):
+	// 		- The application is multi source and required version too -> multi source
+	// 		- The application is single source and the required version too -> single source
+	// 		- The application is multi source and the required version is single source -> single source
+	// 		- The application is single source and the required version is multi source -> multi source
+	// - Historic revision:
+	// 		- The application is multi source and the previous one too -> multi source
+	// 		- The application is single source and the previous one too -> single source
+	// 		- The application is multi source and the previous one is single source -> multi source
+	// 		- The application is single source and the previous one is multi source -> single source
+	isRevisionMultiSource := a.Spec.HasMultipleSources()
+	emptyHistory := len(a.Status.History) == 0
+	if !emptyHistory {
+		for _, h := range a.Status.History {
+			if h.ID == versionId {
+				isRevisionMultiSource = len(h.Revisions) > 0
+				break
+			}
+		}
+	}
+
+	// If the historical data is empty (because the app hasn't been synced yet)
+	// we can use the source, if not (the app has been synced at least once)
+	// we have to use the history because sources can be added/removed
+	if emptyHistory {
+		if isRevisionMultiSource {
+			source = &a.Spec.Sources[*q.SourceIndex]
+		} else {
+			s := a.Spec.GetSource()
+			source = &s
+		}
+	} else {
+		// the source count can change during the time, we cannot just trust in .status.sync
+		// because if a source has been added/removed, the revisions there won't match
+		// as this is only used for the UI and not internally, we can use the historical data
+		// using the specific revisionId
+		for _, h := range a.Status.History {
+			if h.ID == versionId {
+				h := h
```

Comment:
any reason for the need of re-assigning the value for this variable?

Reply:
Yeah, as source is a pointer, I redefine the var within the scope to hide the iteration one: https://go.dev/blog/loopvar-preview

Reply:
https://go.dev/play/p/fhnEXSoWXlb?v=goprev